go-controller/pkg/network-attach-def-controller/network_attach_def_controller.go

```diff
@@ -280,3 +323,50 @@ func nadNeedsUpdate(oldNAD, newNAD *nettypes.NetworkAttachmentDefinition) bool {
 
 	return !reflect.DeepEqual(oldNAD.Spec, newNAD.Spec)
 }
+
+func (nadController *NetAttachDefinitionController) GetActiveNetworkForNamespace(namespace string) (util.NetInfo, error) {
+	if !util.IsNetworkSegmentationSupportEnabled() {
+		return &util.DefaultNetInfo{}, nil
+	}
+	nadController.RLock()
+	defer nadController.RUnlock()
+	primaryNAD := nadController.primaryNADs[namespace]
+	if primaryNAD != "" {
+		// we have a primary NAD, get the network
+		netName := nadController.nads[primaryNAD]
+		if netName == "" {
+			// this should never happen where we have a nad keyed in the primaryNADs
+			// map, but it doesn't exist in the nads map
+			panic("NAD Controller broken consistency between primary NADs and cached NADs")
+		}
+		network := nadController.networkManager.getNetwork(netName)
+		networkNADs := network.GetNADs()
+		filteredNADs := make([]string, 0, len(networkNADs))
+		for _, nadName := range networkNADs {
+			nadNs, _, err := cache.SplitMetaNamespaceKey(nadName)
+			if err != nil {
+				klog.Errorf("Cannot parse NAD reference: %q on network info: %q", nadName, netName)
+				continue
+			}
+			if nadNs == namespace {
+				filteredNADs = append(filteredNADs, nadName)
+			}
+		}
+		n := util.CopyNetInfo(network)
+		n.SetNADs(filteredNADs...)
+		return n, nil
+	}
+
+	// no primary network found, make sure we just haven't processed it yet and no UDN exists
+	udns, err := nadController.udnLister.UserDefinedNetworks(namespace).List(labels.Everything())
+	if err != nil {
+		return nil, fmt.Errorf("error getting user defined networks: %w", err)
+	}
+	for _, udn := range udns {
+		if util.IsPrimaryNetwork(udn.Spec) {
+			return nil, util.NewUnprocessedActiveNetworkError(namespace, udn.Name)
+		}
+	}
```

Comment:
food for thought: we can have a controller listening to UDNs so that we can have a map of primary UDNs to make this also faster. Because, aren't we going to end up doing this search always for the default network?

Reply:
agree. that could be a follow up PR. I'll see if its something easy i could add to this PR.

Reply:
lets do this as follow up i dont want to delay this PR any longer